src/nix/app.cc

```diff
@@ -65,6 +65,15 @@ UnresolvedApp Installable::toApp(EvalState & state)
 
     auto type = cursor->getAttr("type")->getString();
 
+    std::string expected;
+    if (hasPrefix(attrPath,"apps.")) {
+        expected = "app";
+    } else {
+        expected = "derivation";
+    }
+    if (type != expected) {
+        throw Error("Attribute '%s' should have type '%s'. Apps must be located in the apps output attribute. The apps output attribute must contain only apps.", attrPath, expected);
+    }
```

Comment:
I don’t find the error message very clear. I’d suggest to either stay very generic (`Attribute '%s' should have type '%s'`) or have a totally different error messages in both cases.

Reply:
I was trying to be explicit that:

>  Attributes shall be of type 'app' if and only if they are in the apps output schema in order to be correctly handled by `nix run` and `nix flake check`

Not sure of the best way to specify this.... reverting.....